### ISSUE TITLE
Okta Authenticator: Fail to extract SAML from html

### ISSUE THREAD

**Issue by ken-ky-wong** [snowflakedb/snowflake-connector-net#602]
### Issue description
The connector has been working with version 2.0.13, until Okta was upgraded last month, and now using Okta Identity Engine.  Since the upgrade the connector throws error "Fail to extract SAML from html", even if I try version 2.0.20.

In `OktaAuthenticator.cs`, function `VerifyPostbackUrl()` around line 185, `samlRawHtmlString.Substring(startIndex, length)` returns `/login/cert`.

I checked with our infrastructure team and have been assured that the Okta user account does not require MFA.

### Example code
```cs
using (IDbConnection conn = new SnowflakeDbConnection())
{
	conn.ConnectionString =
		"account=***.ap-southeast-2;authenticator=https://***.okta.com;user=***;warehouse=***;role=***;db=***;schema=***;password=***;INSECUREMODE=true;CONNECTION_TIMEOUT=120;";
	conn.Open();
	logger.LogInformation("Snowflake connection successful");
	
	// the actual query cmd etc from here
}

```

### Error log
```
[16:37:46 DBG] Open Connection.
[16:37:46 DBG] SnowflakeDbConnectionPool::getConnection
[16:37:46 DBG] SnowflakeDbConnectionPool::initConnectionPool
[16:37:46 INF] Start parsing connection string.
[16:37:46 DBG] Sesssion property PORT set to default value: 443
[16:37:46 DBG] Sesssion property SCHEME set to default value: https
[16:37:46 DBG] Sesssion property VALIDATE_DEFAULT_PARAMETERS set to default value: true
[16:37:46 DBG] Sesssion property USEPROXY set to default value: false
[16:37:46 DBG] Sesssion property DISABLERETRY set to default value: false
[16:37:46 DBG] Sesssion property FORCERETRYON404 set to default value: false
[16:37:46 DBG] Sesssion property CLIENT_SESSION_KEEP_ALIVE set to default value: false
[16:37:46 INF] Compose host name: ***.ap-southeast-2.snowflakecomputing.com
[16:37:46 DBG] Http client for False;;;;;;False;False not registered. Adding.
[16:37:46 DBG] Open Session
[16:37:46 INF] Okta Authentication
[16:37:46 DBG] step 1: get sso and token url
[16:38:24 DBG] Executing: POST https://***.ap-southeast-2.snowflakecomputing.com/session/authenticator-request HTTP/1.1
[16:38:24 DBG] Http request timeout : 00:00:16
[16:38:24 DBG] Rest request timeout : 00:02:00
[16:38:24 DBG] Success Response: StatusCode: 200, ReasonPhrase: 'OK'
[16:38:27 DBG] step 2: verify urls fetched from step 1
[16:38:27 DBG] Checking sso url
[16:38:27 DBG] Checking token url
[16:38:27 DBG] step 3: get idp onetime token
[16:38:32 DBG] Executing: POST https://***.okta.com/api/v1/sessions?additionalFields=cookieToken HTTP/1.1
[16:38:32 DBG] Http request timeout : 00:00:16
[16:38:32 DBG] Rest request timeout : 00:02:00
[16:38:36 DBG] Success Response: StatusCode: 200, ReasonPhrase: 'OK'
[16:38:38 DBG] step 4: get SAML reponse from sso
[16:38:43 DBG] Executing: GET https://***.okta.com/app/snowflake/exk800eg1nnCkTCLD1t7/sso/saml?RelayState=%2Fsome%2Fdeep%2Flink&onetimetoken=**** HTTP/1.1
[16:38:43 DBG] Http request timeout : -00:00:00.0010000
[16:38:43 DBG] Rest request timeout : 00:02:00
[16:38:44 DBG] Success Response: StatusCode: 200, ReasonPhrase: 'OK'
[16:38:44 DBG] step 5: verify postback url in SAML reponse
[16:39:26 ERR] Fail to extract SAML from html
System.UriFormatException: Invalid URI: The format of the URI could not be determined.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind, UriCreationOptions& creationOptions)
   at System.Uri..ctor(String uriString)
   at Snowflake.Data.Core.Authenticator.OktaAuthenticator.VerifyPostbackUrl() in C:\***\snowflake-connector-net\Snowflake.Data\Core\Authenticator\OktaAuthenticator.cs:line 188
[16:39:26 ERR] Unable to connect
Snowflake.Data.Client.SnowflakeDbException (0x80004005): Error: Cannot found the postback url from the SAML response SqlState: , VendorCode: 270041, QueryId: 
 ---> System.UriFormatException: Invalid URI: The format of the URI could not be determined.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind, UriCreationOptions& creationOptions)
   at System.Uri..ctor(String uriString)
   at Snowflake.Data.Core.Authenticator.OktaAuthenticator.VerifyPostbackUrl() in C:\***\snowflake-connector-net\Snowflake.Data\Core\Authenticator\OktaAuthenticator.cs:line 188
   at Snowflake.Data.Core.Authenticator.OktaAuthenticator.VerifyPostbackUrl() in C:\***\snowflake-connector-net\Snowflake.Data\Core\Authenticator\OktaAuthenticator.cs:line 192
   at Snowflake.Data.Core.Authenticator.OktaAuthenticator.Snowflake.Data.Core.Authenticator.IAuthenticator.Authenticate() in C:\***\snowflake-connector-net\Snowflake.Data\Core\Authenticator\OktaAuthenticator.cs:line 112
   at Snowflake.Data.Core.SFSession.Open() in C:\***\snowflake-connector-net\Snowflake.Data\Core\SFSession.cs:line 243
   at Snowflake.Data.Client.SnowflakeDbConnection.Open() in C:\***\snowflake-connector-net\Snowflake.Data\Client\SnowflakeDbConnection.cs:line 203
[16:39:26 DBG] Close Connection.
[16:39:26 DBG] SnowflakeDbConnectionPool::addConnection
[16:39:26 ERR] Failed to load adviser data from Snowflake.
Snowflake.Data.Client.SnowflakeDbException (0x80004005): Error: Cannot found the postback url from the SAML response SqlState: , VendorCode: 270041, QueryId: 
 ---> System.UriFormatException: Invalid URI: The format of the URI could not be determined.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind, UriCreationOptions& creationOptions)
   at System.Uri..ctor(String uriString)
   at Snowflake.Data.Core.Authenticator.OktaAuthenticator.VerifyPostbackUrl() in C:\***\snowflake-connector-net\Snowflake.Data\Core\Authenticator\OktaAuthenticator.cs:line 188
   at Snowflake.Data.Core.Authenticator.OktaAuthenticator.VerifyPostbackUrl() in C:\***\snowflake-connector-net\Snowflake.Data\Core\Authenticator\OktaAuthenticator.cs:line 192
   at Snowflake.Data.Core.Authenticator.OktaAuthenticator.Snowflake.Data.Core.Authenticator.IAuthenticator.Authenticate() in C:\***\snowflake-connector-net\Snowflake.Data\Core\Authenticator\OktaAuthenticator.cs:line 112
   at Snowflake.Data.Core.SFSession.Open() in C:\***\snowflake-connector-net\Snowflake.Data\Core\SFSession.cs:line 243
   at Snowflake.Data.Client.SnowflakeDbConnection.Open() in C:\***\snowflake-connector-net\Snowflake.Data\Client\SnowflakeDbConnection.cs:line 203
```

### Configuration
*Driver version:* 2.0.13, 2.0.20

*Dotnet framework and version:* .net 6

*Server version:* 7.5.1

*Client OS:* Windows 10

**Comment by ken-ky-wong** [comment]
In trying to verify the Okta account is fine, I used the JDBC driver (v3.13.27) to test the connection.  The connection string is almost identical:

`jdbc:snowflake://***.snowflakecomputing.com?authenticator=https://***.okta.com&user=***&warehouse=***&role=***&db=***&schema=***&password=***&INSECUREMODE=true&CONNECTION_TIMEOUT=120`

and I got a successful connection and could do queries.

**Comment by ken-ky-wong** [comment]
With the JDBC driver working, we can observe that the JDBC Authenticator requests pass with headers "CLIENT_APP_ID" and "CLIENT_APP_VERSION".  The "CLIENT_APP_VERSION" seems to be the key to inform the server what type of response to return.

The Authenticator requests for this .net connection do not send those headers and therefore the server always send response the old way.  Tried changing OktaAuthenticator.BuildAuthenticatorRestRequest() to send those headers and seem to have done the trick. 

I guess a PR is needed.

**Comment by ken-ky-wong** [comment]
#603

**Comment by sfc-gh-igarish** [comment]
Merged the PR. Fix will be available in June release.

**Comment by sfc-gh-dszmolka** [comment]
fix is released with version 2.0.25, closing this issue
